### PR TITLE
network: change connection address methods to return pointers

### PR DESCRIPTION
**Pull request by mattklein123**
This is a cleanup needed for further access log work. There is no
functional change.

*Risk Level*: Low 
*Testing*: Existing tests
*Docs Changes*: N/A
*Release Notes*: N/A

### CODE REVIEW

File: tools/path_fix.sh [hunk 1]
```diff
@@ -7,14 +7,14 @@
 #
 # NOTE: This implementation is far from perfect and will need to be refined to cover all cases.
 
-set -e
```
Comment (alyssawilk) [inline tools/path_fix.sh]: Why do we want to lose the fast-exit?

File: tools/path_fix.sh [hunk 2]
```diff
@@ -7,14 +7,14 @@
...
-
 $* 2>&1 |
   while IFS= read -r LINE
   do
     if [[ "${LINE}" =~ [[:space:]]*([^:[:space:]]+):[[:digit:]]+:[[:digit:]]+: ]]; then
       REAL_PATH=$(readlink -f "${BASH_REMATCH[1]}")
-      LINE=${LINE//${BASH_REMATCH[1]}/${REAL_PATH}}
+      if [[ $? == 0 ]]; then
```
Comment (alyssawilk) [inline tools/path_fix.sh]: Bonus points if you comment what this is doing for the folks who can't read it as easily :-P

File: include/envoy/network/connection.h [hunk 3]
```diff
@@ -151,15 +151,15 @@ class Connection : public Event::DeferredDeletable, public FilterManager {
   /**
    * @return The address of the remote client.
    */
-  virtual const Address::Instance& remoteAddress() const PURE;
+  virtual Network::Address::InstanceConstSharedPtr remoteAddress() const PURE;
```
Comment (ggreenway) [inline include/envoy/network/connection.h]: I think these should be documented that they will never return nullptr (unless you plan to change that)

File: tools/path_fix.sh (hunk 1 above)
Comment (ggreenway) [inline tools/path_fix.sh]: I don't feel strongly about this, but shouldn't this tool fix be a separate PR?  Does this have anything to do with the change?

**Comment by ggreenway** [comment]
I'm +1 on the overall change, modulo my two comments

File: tools/path_fix.sh (hunk 1 above)
Comment (mattklein123) [reply tools/path_fix.sh]: I forgot this was in here. I will do it in a different PR with more comments.

**Comment by alyssawilk** [comment]
oops - looks like E_INSUFFICENT_ReturnRef

**Comment by mattklein123** [comment]
yeah I'm fixing, sorry pushed too soon.

**Comment by mattklein123** [comment]
This is still broken. I will reopen when fixed.

**Comment by mattklein123** [comment]
@alyssawilk @ggreenway this should be fixed now. Changing to const ref return actually turned into a huge pain, but, I was able to clean up a bunch of mock usage in the tests.

File: test/common/filter/tcp_proxy_test.cc [hunk 4]
```diff
@@ -467,7 +434,7 @@ class TcpProxyTest : public testing::Test {
                 factory_context_.cluster_manager_.thread_local_cluster_.cluster_.info_));
         ON_CALL(*upstream_hosts_.at(i), address()).WillByDefault(Return(upstream_remote_address_));
         ON_CALL(*upstream_connections_.at(i), localAddress())
-            .WillByDefault(Return(upstream_local_address_));
+            .WillByDefault(ReturnRef(upstream_local_address_));
```
Comment (alyssawilk) [inline test/common/filter/tcp_proxy_test.cc]: Should we remove the local variable here and just use the mock defaults?

File: test/common/filter/tcp_proxy_test.cc [hunk 5]
```diff
@@ -824,7 +791,8 @@ TEST_F(TcpProxyTest, AccessLogUpstreamLocalAddress) {
 TEST_F(TcpProxyTest, AccessLogDownstreamAddress) {
   Network::Address::InstanceConstSharedPtr downstream_address =
       Network::Utility::resolveUrl("tcp://1.1.1.1:40000");
-  ON_CALL(filter_callbacks_.connection_, remoteAddress()).WillByDefault(Return(downstream_address));
+  ON_CALL(filter_callbacks_.connection_, remoteAddress())
+      .WillByDefault(ReturnRef(downstream_address));
```
Comment (alyssawilk) [inline test/common/filter/tcp_proxy_test.cc]: Ditto here.
And would it make sense to have this be the default behavior or would that cause problems for other tests?

File: test/common/http/conn_manager_impl_test.cc [hunk 6]
```diff
@@ -95,7 +95,8 @@ class HttpConnectionManagerImplTest : public Test, public ConnectionManagerConfi
     ON_CALL(filter_callbacks_.connection_, ssl()).WillByDefault(Return(ssl_connection_.get()));
     ON_CALL(Const(filter_callbacks_.connection_), ssl())
         .WillByDefault(Return(ssl_connection_.get()));
-    ON_CALL(filter_callbacks_.connection_, remoteAddress()).WillByDefault(Return(remote_address_));
+    ON_CALL(filter_callbacks_.connection_, remoteAddress())
+        .WillByDefault(ReturnRef(remote_address_));
```
Comment (alyssawilk) [inline test/common/http/conn_manager_impl_test.cc]: Ditto here, and pretty much for all ON_CALLs *if* it works.

**Comment by mattklein123** [comment]
@alyssawilk updated
